Expose pending payments through `ChannelManager`

jurvis opened the pull request:
Addresses https://github.com/lightningdevkit/rust-lightning/issues/1157

jurvis commented:
Still need to add tests and documentation, but seeking concept ACK

jurvis commented:
Tests currently only make assertions for `PendingPaymentStatus::Fulfilled` and `PendingPaymentStatus::Retryable`.

I was thinking of testing `PendingPaymentStatus::Abandoned` by making an assertion after `abandon_payment` is called in this [test](https://github.com/jurvis/rust-lightning/blob/jurvis/2022-11-expose-pending-payments/lightning/src/ln/payment_tests.rs#L1235-L1236), but it looks like it gets removed from `pending_outbound_payments` since its `session_privs.len()` is `0`. Any suggestions on how to proceed?

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1203,6 +1204,36 @@ impl ChannelDetails {
+/// and they can take on either one of the following states as defined in this enum.
+#[derive(Debug, PartialEq)]
+pub enum PendingPaymentStatus {
+	/// When a payment is still being sent and awaiting successful delivery.
```
I copied the docs for the enum cases here basically verbatim from [`PendingOutboundPayment`](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/ln/channelmanager.rs#L473). Let me know if there is a better/preferred way of handling this.

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1203,6 +1204,36 @@ impl ChannelDetails {
+#[derive(Debug, PartialEq)]
+pub enum PendingPaymentStatus {
+	/// When a payment is still being sent and awaiting successful delivery.
+	Retryable,
```
I think it's potentially confusing to call this case `Pending` since looking up the enum `PendingPaymentStatus` from the user's perspective already hints implicitly that these cases are meant to describe _how_ the payments are pending. 

I don't have an issue with calling this `Retryable` mostly because it suitably describes the complement of `Fulfilled` and `Abandoned, which express a "success but pending confirmation" and "user has explicitly given up, or payer will not longer _retry_" states respectively. thoughts?

jkczyz replied:
Brainstorming:
- what do we want to mean by "pending" in the enum type?
- is there another word that better describes what we'd like to convey (tracked? outstanding? revokable? unresolved?) such that pending could make sense for this variant?

jkczyz replied:
Yeah, agreed. The alternatives that I mentioned were more for the top-level type, letting us use "pending" for the variant. Not sold on "in-flight" given we'd overload the meaning. Indifferent to "recent" as I suppose it conveys what is represented but not specifically on what is defined to be "recent" without explanatory docs.

jurvis replied:
resolved in fb0b1afc

jurvis commented:
@TheBlueMatt thanks for the review and feedback on the documentation. My apologies that it has taken me some time to address the comments. 

I just pushed [fb0b1af](https://github.com/lightningdevkit/rust-lightning/pull/1873/commits/fb0b1afc0d2c20224c7603759408a5c05ac8a089) that tidied up the language for users and renamed the `PendingPaymentStatus` enum to `PendingPaymentDetails` instead after moving `amount_msat` into the `Retryable` case.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1203,6 +1204,26 @@ impl ChannelDetails {
+		/// inflight.
+		total_msat: u64
+	},
+	/// When a pending payment is fulfilled, we continue tracking it until all pending HTLCs have
+	/// been resolved. These payments will only cease being tracked upon receiving or handling
+	/// [`Event::PaymentSent`].
+	Fulfilled,
+	/// After a payment is explicitly abandoned by calling [`ChannelManager::abandon_payment`], they
+	/// are marked as abandoned until an [`Event::PaymentFailed`] event is generated. A payment
+	/// could also be marked as abandoned if pathfinding fails repeatedly and we will no longer
+	/// retry sending.
+	Abandoned,
```
How useful are these without any other information? Should we at least include the payment hash in all variants?

jurvis replied:
good point. I added it in 7efad83b

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1203,6 +1204,35 @@ impl ChannelDetails {
+	/// After a payment is explicitly abandoned by calling [`ChannelManager::abandon_payment`], they
+	/// are marked as abandoned until an [`Event::PaymentFailed`] event is generated. A payment
+	/// could also be marked as abandoned if pathfinding fails repeatedly and we will no longer
+	/// retry sending.
```
Resolved in b4294ec9

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1203,6 +1204,35 @@ impl ChannelDetails {
+		payment_hash: Option<PaymentHash>,
+	},
+	/// After a payment is explicitly abandoned by calling [`ChannelManager::abandon_payment`], they
+	/// are marked as abandoned until an [`Event::PaymentFailed`] event is generated. A payment
```
Resolved in b4294ec9

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1203,6 +1204,35 @@ impl ChannelDetails {
+	},
+	/// When a pending payment is fulfilled, we continue tracking it until all pending HTLCs have
+	/// been resolved. These payments will only cease being tracked upon receiving or handling
+	/// [`Event::PaymentSent`].
```
I added a mention of a "delay" in b4294ec9 but am not sure if this is necessarily the best approach. However, seems fair enough given that `IDEMPOTENCY_TIMEOUT_TICKS` is private and we probably do not want to hard-code the delay value (7) here.

jkczyz replied:
Actually, should be a few _minutes_. IIUC, `timer_tick_occurred` should be called roughly every minute.

jurvis replied:
oh, makes sense.

jurvis replied:
resolved in 9cb1d121

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1203,6 +1204,35 @@ impl ChannelDetails {
 	}
 }
 
+/// Used to express the status of payments that are currently pending resolution.
```
resolved in b4294ec9

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1770,6 +1800,27 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
+						total_msat: *total_msat,
+					})
+				},
+				PendingOutboundPayment::Abandoned { payment_hash, .. } => {
+					Some(RecentPaymentDetails::Abandoned { payment_hash: *payment_hash })
+				},
+				PendingOutboundPayment::Fulfilled { payment_hash, .. } => {
+					Some(RecentPaymentDetails::Fulfilled { payment_hash: *payment_hash })
+				},
```
bump @TheBlueMatt

jurvis replied:
thanks for clarifying @TheBlueMatt. I thought this was worth clarifying in docs too so I added it in d6266539 for `ChannelManager::list_pending_payments`. Let me know if the wording can be improved/if I missed anything

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1217,16 +1219,16 @@ pub enum RecentPaymentDetails {
-	/// could also be marked as abandoned if pathfinding fails repeatedly and we will no longer
-	/// retry sending.
+	/// After a payment is explicitly abandoned by calling [`ChannelManager::abandon_payment`], it
+	/// is marked as abandoned until an [`Event::PaymentFailed`] event is generated. A payment
```
drop redundant "event"

jurvis replied:
Fixed in 93bf16b0

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1203,6 +1204,37 @@ impl ChannelDetails {
 	}
 }
 
+/// Used to express the status of payments that are currently pending resolution. They do not
```
Let's link to `ChannelManager::list_pending_payments`.

jurvis replied:
Resolved in 61d4e31b

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1203,6 +1204,37 @@ impl ChannelDetails {
+	/// been resolved. Upon receiving [`Event::PaymentSent`], we delay for a couple of seconds
+	/// before the payment is removed from tracking.
+	Fulfilled {
+		/// Hash of the payment that was claimed.
```
Let's say when this would be `None`. Likely for older serializations.

jurvis replied:
Fixed in 00fc748e

jkczyz commented:
Note that #1923 moved `PendingOutboundPayment`, so will need to adjust once we're ready to rebase.

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1204,7 +1204,9 @@ impl ChannelDetails {
-/// Used to express the status of payments that are currently pending resolution.
+/// Used to express the status of payments that are currently pending resolution. They do not
+/// include only payments that are currently inflight, but also those who may have yet to find a
+/// path or have outstanding HTLCs.
```
Resolved in 61d4e31b

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1217,16 +1219,16 @@ pub enum RecentPaymentDetails {
 	/// When a pending payment is fulfilled, we continue tracking it until all pending HTLCs have
-	/// been resolved. These payments will only cease being tracked upon receiving or handling
-	/// [`Event::PaymentSent`].
+	/// been resolved. Upon receiving [`Event::PaymentSent`], we delay for a couple of seconds
```
resolved in 9cb1d121

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1610,6 +1641,32 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
+	/// payment is returned as [`RecentPaymentDetails::Fulfilled`], you may retry it, but should
+	/// not.
+	pub fn list_pending_payments(&self) -> Vec<RecentPaymentDetails> {
+		self.pending_outbound_payments.pending_outbound_payments.lock().unwrap().iter()
```
I think the `pending_outbound_payments` field in `OutboundPayments` may be worth renaming to avoid `self.pending_outbound_payments.pending_outbound_payments`.

thoughts? cc // @jkczyz @valentinewallace @TheBlueMatt

jkczyz replied:
Yeah, I would be supportive of reaming. Perhaps `id_to_pending_payment`, `payment_by_id`, or the like? Open to other alternatives given we may want to be deliberate about where "pending" is used.

jurvis commented:
rebased to fix conflicts from `PendingOutboundPayment` refactor. left a tiny comment on naming that needs some input: https://github.com/lightningdevkit/rust-lightning/pull/1873/files#r1056876283ç

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1122,7 +1122,8 @@ pub enum RecentPaymentDetails {
 	/// payment is removed from tracking.
 	Fulfilled {
-		/// Hash of the payment that was claimed.
+		/// Hash of the payment that was claimed. Returns `None` for possibly older serializations
```
Drop "Returns"

jurvis replied:
resolved in 7ce4291e

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1122,7 +1122,8 @@ pub enum RecentPaymentDetails {
 	/// payment is removed from tracking.
 	Fulfilled {
-		/// Hash of the payment that was claimed.
+		/// Hash of the payment that was claimed. Returns `None` for possibly older serializations
```
Let's be specific. Looks like the optional payment hash was introduced in 0.0.104 (https://github.com/lightningdevkit/rust-lightning/commit/2c4f16d5e39e6d4e90d1b9836cceaad11810fbaa).

jurvis replied:
resolved in 7ce4291e

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1104,6 +1104,37 @@ impl ChannelDetails {
+	Fulfilled {
+		/// Hash of the payment that was claimed. Returns `None` for possibly older serializations
+		/// of [`ChannelManager`]
+		payment_hash: Option<PaymentHash>,
```
@TheBlueMatt Thoughts on removing the `Option` here and in `PendingOutboundPayment`? Looks like we introduced it in `0.0.104` (https://github.com/lightningdevkit/rust-lightning/commit/2c4f16d5e39e6d4e90d1b9836cceaad11810fbaa).

jurvis commented:
back after a bunch of travel, hope to land this soon 🤞

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1104,6 +1104,37 @@ impl ChannelDetails {
 
+/// Used by [`ChannelManager::list_pending_payments`] to express the status of payments that are
+/// currently pending resolution. These not only include payments that are currently inflight, but
+/// also those that may have yet to find a successful path or have outstanding HTLCs.
```
How is "outstanding" defined here? I can see it being read as "in-flight" but that would be redundant with the first part of the sentence. More generally, we should be clear and consistent throughout all the docs and identifiers with these terms, possibly eliminating redundant ones:
- pending
- resolved / unresolved / resolution
- inflight
- outstanding
- tracked
- recent
- fulfilled
- delivery

Both related to HTLCs and payments.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1610,6 +1641,32 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
+	/// Gets a list of payments in random order that are currently pending resolution.
+	///
+	/// This can be useful for figuring out whether or not a payment made needs to be retried.
+	/// In general, if it is not listed here, you should consider retrying it with a caveat: if a
+	/// payment is returned as [`RecentPaymentDetails::Fulfilled`], you may retry it, but should
+	/// not.
```
Would it be simpler to say to "should consider retrying if not [`RecentPaymentDetails::Fulfilled`]"?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1105,8 +1105,8 @@ impl ChannelDetails {
 /// Used by [`ChannelManager::list_pending_payments`] to express the status of payments that are
-/// currently pending resolution. These not only include payments that are currently inflight, but
-/// also those that may have yet to find a successful path or have outstanding HTLCs.
+/// currently pending. These not only include payments that are currently inflight, but also those
```
Not sure if we want to use "pending" here given that is one of the variants. Probably can just say "status of recent payment" given the method rename.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1105,8 +1105,8 @@ impl ChannelDetails {
 }
 
 /// Used by [`ChannelManager::list_pending_payments`] to express the status of payments that are
```
list_recent_payments

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1105,8 +1105,8 @@ impl ChannelDetails {
 /// Used by [`ChannelManager::list_pending_payments`] to express the status of payments that are
-/// currently pending resolution. These not only include payments that are currently inflight, but
-/// also those that may have yet to find a successful path or have outstanding HTLCs.
+/// currently pending. These not only include payments that are currently inflight, but also those
+/// that may have yet to find a successful path or have unresolved HTLCs.
```
Seems that "payments that are currently inflight" is defined by what follows, so saying "not only" and "but also" isn't accurate. What suggest dropping those words and replacing the later simply with "i.e.,".

jurvis replied:
I just made it all one line: "These include payments that have yet to find a successful path, or have unresolved HTLCs."

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1668,6 +1698,31 @@ where
 		self.list_channels_with_filter(|&(_, ref channel)| channel.is_live())
 	}
 
+	/// Gets a list of payments in random order that are currently pending resolution.
```
Similar here regarding pending resolution. Can take from `RecentPaymentDetails` docs and also:

s/Gets a list of/Returns recent
s/random/an undefined

jkczyz reviewed (commented):
Generally LGTM. Just need to match some docs in `list_recent_payments`.

jkczyz reviewed (commented):
Just some grammar nits

jkczyz commented:
Regarding the failed CI checks, let's see if a rebase helps once @TheBlueMatt is ready.

jurvis commented:
hmm, looks like there is still a lockorder violation, likely due to accessing `pending_outbound_payments` -- any ideas?

jkczyz commented:
> hmm, looks like there is still a lockorder violation, likely due to accessing `pending_outbound_payments` -- any ideas?

`get_channel_ref!` macro takes some locks, so you'll want to move the calls to `list_recent_payments` outside of the scope in `test_trivial_inflight_htlc_tracking`.

jurvis commented:
ah, gotcha. that was it. thanks for the help!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1668,6 +1698,32 @@ where
+	/// successful path, or have unresolved HTLCs.
+	///
+	/// This can be useful for figuring out whether or not a payment needs to be retried.
+	/// In general, if it is not listed here, you should consider retrying it if not
```
Hmm... yeah, I think we really need a precise description of the problem to include here, or possibly even somewhere else. As a user, I'd have to come across these docs by happenstance to even know that I need to be aware of this. Seems like at very least we need to incorporate this into `ChannelManager::send_payment`?

jurvis replied:
@TheBlueMatt @jkczyz thanks for the feedback! I will go ahead and adjust the language here and also at `send_payment` so users can more clearly see the motivation behind this API

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1701,9 +1701,11 @@ where
-	/// In general, if it is not listed here, you should consider retrying it if not
-	/// [`RecentPaymentDetails::Fulfilled`].
+	/// This can be useful for payments that may have been prepared, but ultimately not sent, as a
+	/// result of a crash. If such a payment exists, is not listed here, and an
+	/// [`Event::PaymentSent`] event has yet to be received, you may consider retrying the payment.
```
No need to repeat "event". How about "an [`Event::PaymentSent`] has not been received"?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2482,6 +2484,11 @@ where
+	/// Additionally,  if the process of sending a payment begins, but we crash before send_payment
+	/// returns (or prior to MonitorUpdate completion if you're using
+	/// [`ChannelMonitorUpdateStatus::InProgress`]), the payment may be lost on restart. See
+	/// [`ChannelManager::list_recent_payments`] for more information.
```
Actually, what I meant was move what you wrote to that paragraph above rather than repeat part of that down here.

alecchendev reviewed (commented):
Left some comments mainly looking at the public docs from a pair of fresh eyes and noted anything that came to mind :)

jurvis commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1131,6 +1131,36 @@ impl ChannelDetails {
+		/// Hash of the payment that is currently being sent but has yet to be fulfilled or
+		/// abandoned.
+		payment_hash: PaymentHash,
+		/// Total amount (in msat) across all paths for this payment, not just the amount currently
+		/// inflight.
+		total_msat: u64,
```
> When it says "all paths" is that said to account for MPP? Like if there were 2 paths for this payment, total_msat is the sum of the value received across both those paths?

Yup!

> Also I think it would maybe be helpful to add that this doesn't include fees? May need to double check, but from what I see in PendingOutboundPayment::Retryable it doesn't seem to(?). Since this is a sort of simpler interface on top of PendingOutboundPayment and because it lacks the other fee related fields that PendingOutboundPayment has (that might help a user infer whether total_msat includes fees), I figured I'd ask if it might be helpful to include in docs.

I think that's valid feedback, thanks for pointing it out -- I'll add it :)

jurvis commented:
> Left some comments mainly looking at the public docs from a pair of fresh eyes and noted anything that came to mind :)

@alecchendev love it! thanks for the review and welcome to the project!

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1131,6 +1131,36 @@ impl ChannelDetails {
 	}
 }
 
+/// Used by [`ChannelManager::list_recent_payments`] to express the status of recent payments.
+/// These include payments that have yet to find a successful path, or have unresolved HTLCs.
+#[derive(Debug, PartialEq)]
+pub enum RecentPaymentDetails {
```
Ah true makes sense 👍

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2426,6 +2484,11 @@ where
 	/// irrevocably committed to on our end. In such a case, do NOT retry the payment with a
 	/// different route unless you intend to pay twice!
 	///
+	/// Thus, in order to ensure duplicate payments are not sent, if we begin the process of sending
+	/// a payment, but crash before `send_payment` returns (or prior to [`ChannelMonitorUpdate`]
+	/// persistence if you're using [`ChannelMonitorUpdateStatus::InProgress`]), the payment may be
+	/// lost on restart. See [`ChannelManager::list_recent_payments`] for more information.
+	///
```
Err.. sorry, what I meant was what you wrote in the previous commit belongs as part of the paragraph starting on line 2461. That paragraph says, "you should implement your own tracking of payments." So it seems like a natural place to talk about `list_recent_payments`.

jurvis commented:
Added some doc improvements and rebased to resolve a conflict on [90ce06b](https://github.com/lightningdevkit/rust-lightning/pull/1873/commits/90ce06bbf45782f93b97b9e9e0b619ba2c5dc529)

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2476,6 +2476,8 @@ where
 	/// Value parameters are provided via the last hop in route, see documentation for RouteHop
 	/// fields for more info.
 	///
+	///	# Avoiding Duplicate Payments
```
Space not tab here

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2488,12 +2490,19 @@ where
+	/// using [`ChannelMonitorUpdateStatus::InProgress`]), the payment may be lost on restart. See
+	/// [`ChannelManager::list_recent_payments`] for more information.
+	///
 	/// May generate SendHTLCs message(s) event on success, which should be relayed (e.g. via
 	/// [`PeerManager::process_events`]).
```
Let's move this to the end of the second paragraph (i.e., after "Value parameters are provided via ...").

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2488,12 +2490,19 @@ where
 	/// May generate SendHTLCs message(s) event on success, which should be relayed (e.g. via
 	/// [`PeerManager::process_events`]).
 	///
+	/// # Possible Error States on PaymentSendFailure
```
Add ticks on PaymentSendFailure

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2476,6 +2476,8 @@ where
 	/// Value parameters are provided via the last hop in route, see documentation for RouteHop
```
Tick / link RouteHop

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2508,20 +2517,17 @@ where
+	/// `payment_secret` is unrelated to `payment_hash` (or [`PaymentPreimage`]) and exists to
+	/// authenticate the sender to the recipient and prevent payment-probing (deanonymization)
+	/// attacks. For newer nodes, it will be provided to you in the invoice. If you do not have one,
+	/// the Route must not contain multiple paths as multi-path payments require a
```
Tick / link Route

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2508,20 +2517,17 @@ where
-	/// we assume the invoice had the basic_mpp feature set.
+	/// If a `payment_secret` *is* provided, we assume that the invoice had the payment_secret
+	/// feature bit set (either as required or as available). If multiple paths are present in the
+	/// Route, we assume the invoice had the basic_mpp feature set.
```
Likewise.

jkczyz reviewed (commented):
Thanks for cleaning up the docs! Much easier to parse them with the headings. Just some nits around them.
